ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/ClientTest.java

```diff
@@ -0,0 +1,308 @@
+    assertThat(testEndpoints.getLastProperties().getMap(), is(Collections.emptyMap()));
+
+    assertThat(rows, hasSize(DEFAULT_ROWS.size()));
+    for (int i = 0; i < DEFAULT_ROWS.size(); i++) {
+      assertThat(rows.get(i).values(), equalTo(rowWithIndex(i).getList()));
+      assertThat(rows.get(i).columnNames(), equalTo(DEFAULT_COLUMN_NAMES));
+      assertThat(rows.get(i).columnTypes(), equalTo(DEFAULT_COLUMN_TYPES));
+    }
+
+    assertThat(server.getQueryIDs(), hasSize(0));
+  }
+
+  @Test
+  public void shouldHandleErrorResponseFromExecuteQuery() throws Exception {
+    // Given
+    ParseFailedException pfe = new ParseFailedException("invalid query blah");
+    testEndpoints.setCreateQueryPublisherException(pfe);
+
+    // When/Then
+    assertErrorWhen(
+        () -> client.executeQuery("bad query"),
+        "Received 400 response from server",
+        "invalid query blah"
+    );
+  }
+
+  @Test
+  public void shouldFailExecuteQueryOnPushQuery() throws Exception {
+    assertErrorWhen(
+        () -> client.executeQuery(DEFAULT_PUSH_QUERY),
+        "Expecting pull query but was push query"
+    );
+  }
+
+  protected Client createJavaClient() {
+    return new ClientImpl(createJavaClientOptions(), vertx);
+  }
+
+  protected ClientOptions createJavaClientOptions() {
+    return new ClientOptionsImpl()
+        .setHost("localhost")
+        .setPort(server.getListeners().get(0).getPort())
+        .setUseTls(false);
+  }
+
+  private void verifyPushQueryMetadata(final QueryResult queryResult) {
+    assertThat(testEndpoints.getLastSql(), is(DEFAULT_PUSH_QUERY));
+    assertThat(testEndpoints.getLastProperties(), is(BaseApiTest.DEFAULT_PUSH_QUERY_REQUEST_PROPERTIES));
+
+    assertThat(queryResult.columnNames(), is(DEFAULT_COLUMN_NAMES));
+    assertThat(queryResult.columnTypes(), is(DEFAULT_COLUMN_TYPES));
+
+    String queryId = queryResult.queryID();
+    assertThat(queryId, is(notNullValue()));
+    assertThat(server.getQueryIDs(), hasSize(1));
+    assertThat(server.getQueryIDs().contains(new PushQueryId(queryId)), is(true));
+  }
+
+  private void shouldDeliver(final Publisher<Row> publisher, final int numRows) {
+    TestSubscriber<Row> subscriber = new TestSubscriber<Row>() {
+      @Override
+      public synchronized void onSubscribe(final Subscription sub) {
+        super.onSubscribe(sub);
+        sub.request(numRows);
+      }
+    };
+    publisher.subscribe(subscriber);
+    assertThatEventually(subscriber::getValues, hasSize(numRows));
+    for (int i = 0; i < numRows; i++) {
+      assertThat(subscriber.getValues().get(i).values(), equalTo(rowWithIndex(i).getList()));
+    }
+    assertThat(subscriber.isCompleted(), equalTo(false));
+    assertThat(subscriber.getError(), is(nullValue()));
+  }
+
+  private void assertErrorWhen(
+      final Supplier<CompletableFuture<?>> queryRequest,
+      final String... errorMessages
+  ) throws Exception {
+    // Given
+    CountDownLatch latch = new CountDownLatch(1);
+
```

Comment:
Wouldn't it be simpler just to fall cf.get() rather than using a latch, and handling everything asynchronously?

Reply:
Updated. I do think this is cleaner, though it's a bit odd that the expected exception type is always `ExecutionException` since that's what's thrown when `.get()` is called on a future that completes exceptionally. Added comments inline to avoid confusion.

Reply:
Yeah ExecutionException wrapping can be a pain. I think the reasoning is that any exceptions that are thrown from inside completable future handlers will be wrapped in an ExecutionException, Any exceptions thrown internally from the completablefuture framework itself, are not wrapped in ExecutionException, so it's a way of distinguishing the two kinds of exceptions.